File: python/leetcode/easy/average_of_levels_in_binary_tree.py
# https://leetcode.com/problems/average-of-levels-in-binary-tree/

class Solution:
    def __init__(self):
        self.elements = {}

    def averageOfLevels(self, root: TreeNode) -> List[float]:
        result = []
        level = 0
        self.index_elements(root, level)

        for elements in self.elements.values():
            result.append(sum(elements)/len(elements))

        return result


    def index_elements(self, root: TreeNode, level: int):
        if root is None:
            return

        if level not in self.elements:
            self.elements[level] = []

        self.elements[level].append(root.val)
        level += 1

        self.index_elements(root.left, level)
        self.index_elements(root.right, level)

File: go/leetcode/easy/averageOfLevelsInBinaryTree.go
// https://leetcode.com/problems/average-of-levels-in-binary-tree/

func averageOfLevels(root *TreeNode) []float64 {
    level := 0
    elements := make(map[int][]int)
    result := make([]float64, 0)

    maxLevel := indexElements(root, level, elements)
    for i := 0; i < maxLevel; i++ {
        sum, length := 0, 0
        for _, el := range(elements[i]){
            sum += el
            length++
        }
        average := float64(sum) / float64(length)
        result = append(result, average)
    }
    return result
}

func indexElements(root *TreeNode, level int, elements map[int][]int) int {
    if root == nil {
        return level
    }

    elements[level] = append(elements[level], root.Val)
    level++

    left := indexElements(root.Left, level, elements)
    right := indexElements(root.Right, level, elements)
    if left > right {
        return left
    }
    return right
}
